gh-67041: Allow to distinguish between empty and not defined URI components (#123305)
python/cpython: Doc/library/urllib.parse.rst

The URL parsing functions focus on splitting a URL string into its components,
or on combining URL components into a URL string.

> .. function:: urlparse(urlstring, scheme=None, allow_fragments=True, *, missing_as_none=False)

Reviewer: I am really sorry, that I am contributing so late; but "missing_as_none=False" is confusing and not intuitive at all to me.

Pretty sure, others who have not participated are going to feel the same.

The function signature and term is not giving a signal on what it is meant to be. 

Are you open to new name?

Reviewer: I am adding @gpshead, as one of the active developers in this area, to get his opinion too.

Reviewer 2: I'm fine with this name.  Admittedly I have spent too much time in the past wrangling problems in this library, but the reason it still works for me despite that is that it is a common concept: do you represent the absence of a value distinctly from the base zero/empty version of that type or not?  That is what None is for.  and *missing_as_none* is at least explicit in name to indicate that some values may be None.  I'm not going to call it _pretty_ but it is "understandable enough" for me.  I can't come up with anything that'd be meaningfully better rather than just alternately-understandable.

Author: What are your thoughts about `missing_as_empty` with the opposite semantic? In future, None will be returned for not defined components by default, and you will need to specify `missing_as_none=False` or `missing_as_empty=True` to restore the current behavior.

Reviewer 3: I think that name should really be `missing_as_empty_string` at least (we can’t say just “empty”), meaning _missing parts returned as empty strings_.

What about `use_none=False` which is short and doesn’t try to be self-explanatory, so people need to read the docs?

      separator key, with ``&`` as the default separator.

.. function:: urlunparse(parts, *, keep_empty=False)

   Construct a URL from a tuple as returned by ``urlparse()``. The *parts*
   argument can be any six-item iterable.

   This may result in a slightly different, but equivalent URL, if the
   URL that was parsed originally had unnecessary delimiters (for example,
   a ``?`` with an empty query; the RFC states that these are equivalent).

   If *keep_empty* is true, empty strings are kept in the result (for example,
   a ``?`` for an empty query), only ``None`` components are omitted.

>    This allows to restore the URL that was parsed with option

Reviewer 3: ```suggestion
   This allows rebuilding a URL that was parsed with option
```

   ``missing_as_none=True``.

>    By default, *keep_empty* is true if *parts* is the result of the

Reviewer 2: It is easy to miss this "footnote" about keep_empty not being a simple False default.  I think the function signature above should be more clear that it has a non-trivial default value when not explicitly specified.  something awkward with a fake descriptive name but indicative that people should read further for details keep_empty=_FALSE_UNLESS_PARTS_IS_A_URLSPLIT_RESULT or similar perhaps?

Author: Pydoc will output `urlunparse(components, *, keep_empty=['not specified'])`. Is it fine, or you need more descriptive name?

Reviewer 3: I think gps is requesting the signature in this doc file to be more explicit, not pydoc output.

Author: I changed also the signature in this doc file. See above. It no longer contains `keep_empty=False`.

   .. versionchanged:: 3.12
      Leading WHATWG C0 control and space characters are stripped from the URL.

   .. versionchanged:: next
      Added the *missing_as_none* parameter.

.. _WHATWG spec: https://url.spec.whatwg.org/#concept-basic-url-parser

.. function:: urlunsplit(parts, *, keep_empty=False)

   Combine the elements of a tuple as returned by :func:`urlsplit` into a
   complete URL as a string. The *parts* argument can be any five-item
   iterable.

   This may result in a slightly different, but equivalent URL, if the
   URL that was parsed originally had unnecessary delimiters (for example,
   a ``?`` with an empty query; the RFC states that these are equivalent).

   If *keep_empty* is true, empty strings are kept in the result (for example,
   a ``?`` for an empty query), only ``None`` components are omitted.
   This allows to restore the URL that was parsed with option
   ``missing_as_none=True``.

>    By default, *keep_empty* is true if *parts* is the result of the

Reviewer 2: same comment as above.  make it obvious from the function signature that the default depends on the type of parts.
